### ISSUE TITLE
Failing to set up NLB when it is in another VPC than the EKS cluster

### ISSUE THREAD

**Issue by sgielen** [kubernetes-sigs/aws-load-balancer-controller#2381]
**Describe the bug**
We have a platform setup where the load balancers are in another VPC (the Ingress VPC) than the Kubernetes cluster (the Cluster VPC). The aws-alb-ingress-controller is running with `--aws-vpc-id` set to the **ingress VPC**.

When we create ALBs from Ingresses in the Kubernetes cluster, they are correctly created in the Ingress VPC, pointing correctly at targets within the Cluster VPC.
Relevant issue: https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/1569 (for ALBs)
Relevant PR: https://github.com/kubernetes-sigs/aws-load-balancer-controller/pull/1862 (which fixes this case for ALBs)

Now, I need to create an NLB reachable on fixed private IP addresses inside the Ingress VPC, but with a target inside the Cluster VPC. The NLB should use regular IP routing to reach the target. However, when I create a LoadBalancer Service I'm getting the error `The following targets are not in the target group VPC <ingress VPC>: <nodes in the cluster VPC>`. The target group VPC should be the cluster VPC, not the ingress VPC.

**Steps to reproduce**
- Create two VPCs: "Ingress VPC" and "Cluster VPC", with non-overlapping subnets and routing between the two.
- Set up an EKS cluster in the Cluster VPC, and aws-load-balancer-controller with command-line flag `--aws-vpc-id <id of the Ingress VPC>`.
- Create a Service like this one (I've tried various values for load-balancer-type and nlb-target-type, none of them worked):

```yaml
apiVersion: v1
kind: Service
metadata:
  name: nlb-service
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-type: nlb
    service.beta.kubernetes.io/aws-load-balancer-scheme: internal
    service.beta.kubernetes.io/aws-load-balancer-internal: "true"
    # nlb-ip or external
    #service.beta.kubernetes.io/aws-load-balancer-type: external
    # instance or ip
    service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: instance
    # Have one IP address per AZ subnet
    service.beta.kubernetes.io/aws-load-balancer-cross-zone-load-balancing-enabled: "true"
    service.beta.kubernetes.io/aws-load-balancer-private-ipv4-addresses: "subnet1-ip,subnet2-ip,subnet3-ip"
    service.beta.kubernetes.io/aws-load-balancer-subnets: "ingress-subnet1,ingress-subnet2,ingress-subnet3"
spec:
  type: LoadBalancer
  selector:
    app: my-application
  ports:
  - name: my-port
    port: 1234
```

**Expected outcome**
I expected an NLB to be created in the Ingress VPC, with the three indicated IPs, with a target group pointing at the application in the Cluster VPC. When I telnet to any of the three indicated IPs, I expect my communications to end up at the application.

**Actual outcome**
The NLB is created in the Ingress VPC and the three IPs are correct.

However, target group registration fails.

With `aws-load-balancer-nlb-target-type: instance`, I get the following errors in the aws-load-balancer-controller logs:

```
{"level":"info","ts":1637684529.8391867,"msg":"registering targets","arn":"arn:aws:elasticloadbalancing:eu-west-1:137505996109:targetgroup/k8s-<snip>","targets":[{"AvailabilityZone":null,"Id":"i-03e320<snip>","Port":32400},{"AvailabilityZone":null,"Id":"i-08a842<snip>","Port":32400},{"AvailabilityZone":null,"Id":"i-0b<snip>","Port":32400},{"AvailabilityZone":null,"Id":"i-0c233<snip>","Port":32400}]}
{"level":"error","ts":1637684529.970452,"logger":"controller","msg":"Reconciler error","reconcilerGroup":"elbv2.k8s.aws","reconcilerKind":"TargetGroupBinding","controller":"targetGroupBinding","name":"k8s-<snip>","namespace":"m<snip>v","error":"InvalidTarget: The following targets are not in the target group VPC '<Ingress VPC>': '<instances in the Cluster VPC>'\n\tstatus code: 400, request id: ac0f843a-4a4c-428f-ac00-fc43ff57f786"}
```

With `aws-load-balancer-nlb-target-type: ip`, or with `service.beta.kubernetes.io/aws-load-balancer-type: nlb-ip`, I get this error:

```
{"level":"error","ts":1637684265.6596644,"logger":"controller","msg":"Reconciler error","reconcilerGroup":"elbv2.k8s.aws","reconcilerKind":"TargetGroupBinding","controller":"targetGroupBinding","name":"k8s-<snip>","namespace":"m<snip>v","error":"cannot resolve pod ENI for pods: [<application Pod>]"}
```

This may be expected, as we use Cilium CNI and not the native AWS CNI. However, it seems to be working for ALBs, since all existing target groups use target type IP.

**Environment**

* AWS Load Balancer controller version shoekstra/aws-alb-ingress-controller v2.1.3-48-ge0aed65f-dirty (built from this PR: https://github.com/kubernetes-sigs/aws-load-balancer-controller/pull/1862)
* Kubernetes version v1.20.11-eks-f17b81
* Using EKS (yes/no), if so version? yes, eks.2

**Comment by sgielen** [comment]
Closing for now. I think this was because I was reading documentation from the wrong aws-alb-ingress-controller version. I was reading v2.2, where `service.beta.kubernetes.io/aws-load-balancer-internal: "true"` was deprecated in favor of `service.beta.kubernetes.io/aws-load-balancer-scheme: internal`, so I only used the second annotation. However, I was **using** v2.1, where the second is not supported. Adding in the first annotation at least allowed the NLB to be created within the correct Ingress VPC.

**Comment by sgielen** [comment]
With `aws-load-balancer-internal: "true"` added, the error just returned in the logs, I've updated the description accordingly.

**Comment by sgielen** [comment]
I think `nlb-target-type: ip` should be the option that should work, but from the code I can't figure out yet why it works for ALB's, but not for NLB's.

**Comment by M00nF1sh** [comment]
@sgielen 
To create a NLB with this controller, you should use both
```
service.beta.kubernetes.io/aws-load-balancer-type: external
service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: ip
```
BTW ```service.beta.kubernetes.io/aws-load-balancer-type: nlb-ip``` achieve same effect as above two annotation but it's deprecated.

Seems the root cause is `cannot resolve pod ENI for pods: [<application Pod>]"`, since we don't support `Cilium CNI`.
It won't work for ALB as well, unless you explicitly specified `alb.ingress.kubernetes.io/security-groups` which will disable the worker node SG rule management. We didn't offer a way to do same thing for NLB yet, which is planned for v2.4.0.

Before that, you can address it with a custom patch to this file: https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/10aab918280fe6077918ae905fee34459c6c0ce0/pkg/service/model_build_target_group.go#L390
replace the `tgbNetworking := t.buildTargetGroupBindingNetworking(ctx, targetPort, preserveClientIP, *hc.Port, port.Protocol, defaultSourceRanges)` with `var tgbNetworking  *elbv2model.TargetGroupBindingNetworking`

**Comment by sgielen** [comment]
> Before that, you can address it with a custom patch to this file:

Perfect, I will try that. Thanks a lot!

**Comment by sgielen** [comment]
@M00nF1sh I can confirm that the issue is resolved with [that change](https://github.com/sgielen/aws-load-balancer-controller/commit/50116c0a489a8a666894d0814aa1293498b7701a). The Docker image `shoekstra/aws-alb-ingress-controller:5d8695-nlb-no-network` correctly creates working NLBs with a target in its TG pointing directly towards the Pod IP with the following Service definition:

```
apiVersion: v1
kind: Service
metadata:
  name: aws-loadbalancer-service
  namespace: default
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-type: nlb
    service.beta.kubernetes.io/aws-load-balancer-scheme: internal
    service.beta.kubernetes.io/aws-load-balancer-internal: "true"
    service.beta.kubernetes.io/load-balancer-source-ranges: "0.0.0.0/0"

    service.beta.kubernetes.io/aws-load-balancer-type: external
    service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: ip

    service.beta.kubernetes.io/aws-load-balancer-cross-zone-load-balancing-enabled: "true"
    service.beta.kubernetes.io/aws-load-balancer-private-ipv4-addresses: "<IPs in subnets in ingress VPC>"
    service.beta.kubernetes.io/aws-load-balancer-subnets: "<subnets in ingress VPC>"

    service.beta.kubernetes.io/aws-load-balancer-backend-protocol: http
spec:
  type: LoadBalancer
  selector:
    app: nginx
  ports:
  - name: http
    port: 80
```

**Comment by sgielen** [comment]
I'll close this issue since you say a fix is coming in v2.4.0 :) Do you have any idea when that version may be released? Would it be possible to have an 'interim fix' for this issue in v2.3.1, so that we can go back to using unpatched upstream versions?

**Comment by M00nF1sh** [comment]
@sgielen 
At this point it will be in Q1 2022 and before EKS 1.22 launch. (it depends on how many changes we want to pack into that, which haven't finalized yet).
we cannot have a `interim fix` in v2.3.1 due to this is a breaking change and requires knobs to lets customer to opt-out. :(
